examples/auth.rs

```diff
@@ -1,45 +1,279 @@
-use duchess::{
-    java::util::{ArrayList, HashMap, ListExt, MapExt},
-    prelude::*,
-    Jvm,
+use duchess::java::lang::{Throwable, ThrowableExt};
+use duchess::java::util::{ArrayList, ArrayListExt, HashMap as JavaHashMap, MapExt};
+use std::collections::HashMap;
+
+use duchess::{prelude::*, Global, Jvm, Local};
+
+// XX: should we automatically attach allow(dead_code)?
+#[allow(dead_code)]
+mod java_auth {
+    use duchess::java::lang::Throwable;
+
+    duchess::java_package! {
+        package auth;
+
+        class Authenticated { * }
+        class AuthorizeRequest { * }
+        class HttpAuth { * }
+        class HttpRequest { * }
+        class AuthenticationException { * }
+        class AuthenticationExceptionUnauthenticated { * }
+        class AuthenticationExceptionInvalidSecurityToken { * }
+        class AuthenticationExceptionInvalidSignature { * }
+        class AuthorizationException { * }
+        class AuthorizationExceptionDenied { * }
+    }
+
+    // XX: can be removed when we automatically look through extends/implements
+    unsafe impl duchess::plumbing::Upcast<Throwable> for AuthenticationException {}
+    unsafe impl duchess::plumbing::Upcast<Throwable> for AuthenticationExceptionUnauthenticated {}
+    unsafe impl duchess::plumbing::Upcast<Throwable> for AuthenticationExceptionInvalidSecurityToken {}
+    unsafe impl duchess::plumbing::Upcast<Throwable> for AuthenticationExceptionInvalidSignature {}
+    unsafe impl duchess::plumbing::Upcast<Throwable> for AuthorizationException {}
+    unsafe impl duchess::plumbing::Upcast<Throwable> for AuthorizationExceptionDenied {}
+
+    pub use auth::*;
+}
+
+use java_auth::{
+    AuthenticatedExt, AuthenticationExceptionUnauthenticatedExt, AuthorizationExceptionDeniedExt,
+    HttpAuthExt,
 };
 
-duchess::java_package! {
-    package me.ferris;
+pub struct HttpAuth(Global<java_auth::HttpAuth>);
 
-    class AuthenticateResult { * }
-    class HttpAuth { * }
-    class HttpRequest { * }
+#[derive(Debug)]
+pub struct HttpRequest {
+    pub verb: String,
+    pub path: String,
+    pub body_hash: Vec<u8>,
+    pub params: HashMap<String, Vec<String>>,
+    pub headers: HashMap<String, Vec<String>>,
```

Comment:
I don't love that we are authoring this struct by hand. I wonder if we can add some kind of option to have duchess auto-generate it -- or just do it always for record types. Something to think about.